pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,395 @@
+            start: 0,
+            end: 0,
+            size: 0,
+            buffer: std::ptr::null_mut(),
+            private: std::ptr::null_mut()
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct PvNic {
+    xsk: *mut xsk_socket,
+
+    umem: *mut xsk_umem,
+    buffer: *mut c_void,
+    chunk_size: u32,
+
+    // chunk_pool: *mut Vec<u64>,
+    chunk_pool: Vec<u64>,
+    chunk_pool_idx: i32,
+
+    /* XSK rings */
+    fq: xsk_ring_prod,
+    rx: xsk_ring_cons,
+    cq: xsk_ring_cons,
+    tx: xsk_ring_prod,
+}
+
+impl PvNic {
+    // TODO: Change return type with Option<PvNic> for safety.
+    fn new(chunk_size_: u32, chunk_count_: i32) -> PvNic {
+
+        let xsk_ptr;
+        unsafe {
+            let xsk_layout = Layout::new::<xsk_socket>();
+            xsk_ptr = alloc_zeroed(xsk_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let umem_ptr;
+        unsafe {
+            let umem_layout = Layout::new::<xsk_umem>();
+            umem_ptr = alloc_zeroed(umem_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let fq_ptr;
+        unsafe {
+            let fq_layout = Layout::new::<xsk_ring_prod>();
+            fq_ptr = alloc_zeroed(fq_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let rx_ptr;
+        unsafe {
+            let rx_layout = Layout::new::<xsk_ring_cons>();
+            rx_ptr = alloc_zeroed(rx_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let cq_ptr;
+        unsafe {
+            let cq_layout = Layout::new::<xsk_ring_cons>();
+            cq_ptr = alloc_zeroed(cq_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let tx_ptr;
+        unsafe {
+            let tx_layout = Layout::new::<xsk_ring_prod>();
+            tx_ptr = alloc_zeroed(tx_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let a:PvNic = unsafe {
+            PvNic {
+                xsk: xsk_ptr.cast::<xsk_socket>(),
+                umem: umem_ptr.cast::<xsk_umem>(),   // umem is needed to be dealloc after using packetvisor library.
+                buffer: std::ptr::null_mut(),
+                chunk_size: chunk_size_,
+                chunk_pool: Vec::with_capacity(chunk_count_ as usize),
+                chunk_pool_idx: chunk_count_,
+                fq: std::ptr::read(fq_ptr.cast::<xsk_ring_prod>()),
+                rx: std::ptr::read(rx_ptr.cast::<xsk_ring_cons>()),
+                cq: std::ptr::read(cq_ptr.cast::<xsk_ring_cons>()),
+                tx: std::ptr::read(tx_ptr.cast::<xsk_ring_prod>()),
+                }
+            };
+
+        a
+    }
+}
+
+fn _pv_alloc(nic: &mut PvNic) -> u64 {
+    if nic.chunk_pool_idx > 0 {
+        nic.chunk_pool_idx -= 1;
+        nic.chunk_pool[nic.chunk_pool_idx as usize]
+    } else {
+        return u64::MAX;
```

Comment:
in C, this code has been intended to keep the program working with allocating packet to a fixed place somewhere (it's already risky) in case `nic.chunk_pool_idx` is below 0 (it's an unwanted behavior).
To follow this logic, the magic number should be put somewhere in the end even though this function returns `Result` type.
Is there another good working logic to change the return type with `Result` type?